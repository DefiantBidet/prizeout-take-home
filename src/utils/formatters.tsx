/**
 * Formats a numerical value in cents to a US Dollar String.
 * @param  {number} valueInCents    Number representing the value in cents
 * @return {string}                 String Locale formatted for US Dollar
 * @function
 */
export const formatCentsToDollar = (valueInCents: number): string =>
    (valueInCents / 100).toLocaleString('en-US', {
        currency: 'USD',
        style: 'currency',
    });

/**
 * Formats a numerical value as a signed decimal.
 * @param  {number} percentValue    Number representing percent
 * @return {string}                 String Locale formatted for signed decimal
 * @function
 */
export const formatPercent = (percentValue: number): string =>
    percentValue.toLocaleString('en-US', {
        signDisplay: 'exceptZero',
        style: 'decimal',
    });

/**
 * Calculates savings cost and formats to a US Dollar String.
 * @param  {number} cost    Number representing the cost in cents
 * @param  {number} value   Number representing the value in cents
 * @return {string}         String Locale formatted for US Dollar
 * @function
 */
export const calculateSavings = (cost: number, value: number): string =>
    (value / 100 - cost / 100).toLocaleString('en-US', {
        currency: 'USD',
        style: 'currency',
    });
