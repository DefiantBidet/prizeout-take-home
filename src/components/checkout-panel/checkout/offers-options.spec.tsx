import React from 'react';
import '@testing-library/jest-dom';
import { PrizeoutOffer, PrizeoutOfferValueOptions } from 'Slices/offers-slice';
import { createMockPrizeoutOffer, fireEvent, render } from 'Testing/test-utils';
import { formatCentsToDollar, calculateSavings } from 'Utils/formatters';

import CheckoutOfferOptions from './offers-options';

describe('Test CheckoutOfferOptions Component', () => {
    const activeOffer: PrizeoutOffer = createMockPrizeoutOffer();
    const activeOption: PrizeoutOfferValueOptions = activeOffer.giftcard_list[0];
    const clickSpy = jest.fn();

    const props = {
        activeOffer,
        onSelectOffer: clickSpy,
        selectedOption: activeOption,
    };

    beforeEach(() => {
        clickSpy.mockReset();
    });

    test('Component matches snapshot', () => {
        const { asFragment } = render(<CheckoutOfferOptions {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });

    describe('Options Grid', () => {
        test('Grid Header', () => {
            const { queryByText } = render(<CheckoutOfferOptions {...props} />);
            expect(queryByText('Select Redemption Amount')).toBeInTheDocument();
        });

        test('Displays Buttons', () => {
            render(<CheckoutOfferOptions {...props} />);
            const buttons = document.querySelectorAll('.offer-options__grid__button');
            expect(buttons).toHaveLength(activeOffer.giftcard_list.length);
        });

        test('Clicking Button Calls ClickSpy', async () => {
            const { getByRole } = render(<CheckoutOfferOptions {...props} />);
            const secondButtonOptions = activeOffer.giftcard_list[1];

            expect(clickSpy.mock.calls).toHaveLength(0);

            const button = getByRole('button', { name: formatCentsToDollar(secondButtonOptions.cost_in_cents) });
            fireEvent.click(button);

            expect(clickSpy.mock.calls).toHaveLength(1);
        });
    });

    describe('Options Details', () => {
        test('Displays Redemption Amount label', () => {
            const { queryByText } = render(<CheckoutOfferOptions {...props} />);
            expect(queryByText('Select Redemption Amount')).toBeInTheDocument();
        });

        test('Displays Redemption Amount value', () => {
            render(<CheckoutOfferOptions {...props} />);
            const row: Element = document.querySelector('.offer-options__details').children[0];
            const costValue = formatCentsToDollar(activeOption.value_in_cents);

            expect(row.lastElementChild.innerHTML).toBe(costValue);
        });

        test('Displays Bonus label', () => {
            const { queryByText } = render(<CheckoutOfferOptions {...props} />);
            expect(queryByText('Prizeout Bonus (+2.04%)')).toBeInTheDocument();
        });

        test('Displays Bonus value', () => {
            render(<CheckoutOfferOptions {...props} />);
            const row: Element = document.querySelector('.offer-options__details').children[1];
            const value = calculateSavings(activeOption.cost_in_cents, activeOption.value_in_cents);

            expect(row.lastElementChild.innerHTML).toBe(value);
        });

        test('Displays Value label', () => {
            const { queryByText } = render(<CheckoutOfferOptions {...props} />);
            expect(queryByText('You Get')).toBeInTheDocument();
        });

        test('Displays Value value', () => {
            render(<CheckoutOfferOptions {...props} />);
            const row: Element = document.querySelector('.offer-options__details').children[2];
            const value = formatCentsToDollar(activeOption.value_in_cents);

            expect(row.lastElementChild.innerHTML).toBe(value);
        });
    });
});
