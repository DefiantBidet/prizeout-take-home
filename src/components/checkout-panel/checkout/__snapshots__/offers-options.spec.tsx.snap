// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CheckoutOfferOptions Component Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="offer-options"
  >
    <h4>
      Select Redemption Amount
    </h4>
    <div
      class="offer-options__grid"
    >
      <button
        class="offer-options__grid__button offer-options__grid__button--selected"
        id="offer-0"
        name="offer-0"
        type="button"
      >
        $0.00
      </button>
      <button
        class="offer-options__grid__button"
        id="offer-1"
        name="offer-1"
        type="button"
      >
        $9.80
      </button>
      <button
        class="offer-options__grid__button"
        id="offer-2"
        name="offer-2"
        type="button"
      >
        $19.60
      </button>
      <button
        class="offer-options__grid__button"
        id="offer-3"
        name="offer-3"
        type="button"
      >
        $29.39
      </button>
    </div>
    <div
      class="offer-options__details"
    >
      <div
        class="offer-options__details__row"
      >
        <div
          class="offer-options__details__row--label"
        >
          Redemption Amount
        </div>
        <div
          class="offer-options__details__row--amount"
        >
          $0.00
        </div>
      </div>
      <div
        class="offer-options__details__row-highlight"
      >
        <div
          class="offer-options__details__row--label"
        >
          Prizeout Bonus (+2.04%)
        </div>
        <div
          class="offer-options__details__row--amount"
        >
          $0.00
        </div>
      </div>
      <div
        class="offer-options__details__row"
      >
        <div
          class="offer-options__details__row--label"
        >
          You Get
        </div>
        <div
          class="offer-options__details__row--amount"
        >
          $0.00
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
